### PR TITLE
Do not reload code inside gr.NO_RELOAD context

### LINKED ISSUE #6214: Gradio State object does not work / reset when using `gradio app.py` to watch for changes
### Describe the bug

When using a Gradio State object in an app, e.g. the hangman example, in conjunction with `gradio hangman.py`, the state object does not properly reset when a change is detected in any of the watched files and an auto-reload is triggered.
When the state is used, it will result in a KeyError. 

Hangman source:
As found at https://www.gradio.app/guides/state-in-blocks

Auto reload method:
As described in https://www.gradio.app/guides/developing-faster-with-reload-mode

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

Add the following code from the [hangman example](https://www.gradio.app/guides/state-in-blocks) to a file `hangman.py`:

```
import gradio as gr

secret_word = "gradio"

with gr.Blocks() as demo:    
    used_letters_var = gr.State([])
    with gr.Row() as row:
        with gr.Column():
            input_letter = gr.Textbox(label="Enter letter")
            btn = gr.Button("Guess Letter")
        with gr.Column():
            hangman = gr.Textbox(
                label="Hangman",
                value="_"*len(secret_word)
            )
            used_letters_box = gr.Textbox(label="Used Letters")

    def guess_letter(letter, used_letters):
        used_letters.append(letter)
        answer = "".join([
            (letter if letter in used_letters else "_")
            for letter in secret_word
        ])
        return {
            used_letters_var: used_letters,
            used_letters_box: ", ".join(used_letters),
            hangman: answer
        }
    btn.click(
        guess_letter, 
        [input_letter, used_letters_var],
        [used_letters_var, used_letters_box, hangman]
        )
demo.launch()
```

Then:
1. run `gradio hangman.py` and open the app and type some letters. 
2. Things work fine
3. Change anything in the code, e.g. `"Guess Letter"` -> `"Take a guess"`
4. Try guessing a letter and observe the error

### Screenshot

Irrelevant

### Logs

```shell
`gradio hangman.py`
Watching: '<python_path>\Anaconda3\lib\site-packages\gradio', '<file_path>scripts\personal', '<file_path>scripts\personal'

Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Changes detected in: <file_path>scripts\personal\hangman.py
Traceback (most recent call last):
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 427, in call_prediction
    output = await route_utils.call_process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\route_utils.py", line 232, in call_process_api
    output = await app.get_blocks().process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1495, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1266, in preprocess_data
    processed_input.append(state[input_id])
  File "<python_path>\Anaconda3\lib\site-packages\gradio\state_holder.py", line 46, in __getitem__
    block = self.blocks.blocks[key]
KeyError: 23
Traceback (most recent call last):
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 427, in call_prediction
    output = await route_utils.call_process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\route_utils.py", line 232, in call_process_api
    output = await app.get_blocks().process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1495, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1266, in preprocess_data
    processed_input.append(state[input_id])
  File "<python_path>\Anaconda3\lib\site-packages\gradio\state_holder.py", line 46, in __getitem__
    block = self.blocks.blocks[key]
KeyError: 23

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 472, in process_events
    response = await self.call_prediction(awake_events, batch)
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 436, in call_prediction
    raise Exception(str(error) if show_error else None) from error
Exception: 23
Traceback (most recent call last):
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 427, in call_prediction
    output = await route_utils.call_process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\route_utils.py", line 232, in call_process_api
    output = await app.get_blocks().process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1495, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1266, in preprocess_data
    processed_input.append(state[input_id])
  File "<python_path>\Anaconda3\lib\site-packages\gradio\state_holder.py", line 46, in __getitem__
    block = self.blocks.blocks[key]
KeyError: 23
Traceback (most recent call last):
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 427, in call_prediction
    output = await route_utils.call_process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\route_utils.py", line 232, in call_process_api
    output = await app.get_blocks().process_api(
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1495, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "<python_path>\Anaconda3\lib\site-packages\gradio\blocks.py", line 1266, in preprocess_data
    processed_input.append(state[input_id])
  File "<python_path>\Anaconda3\lib\site-packages\gradio\state_holder.py", line 46, in __getitem__
    block = self.blocks.blocks[key]
KeyError: 23

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 472, in process_events
    response = await self.call_prediction(awake_events, batch)
  File "<python_path>\Anaconda3\lib\site-packages\gradio\queueing.py", line 436, in call_prediction
    raise Exception(str(error) if show_error else None) from error
Exception: 23
```

### System Info

```shell
gradio environment
Gradio Environment Information:
------------------------------
Operating System: Windows
gradio version: 4.0.2
gradio_client version: 0.7.0

------------------------------------------------
gradio dependencies in your environment:

aiofiles: 23.2.1
altair: 5.1.2
fastapi: 0.104.1
ffmpy: 0.3.1
gradio-client==0.7.0 is not installed.
httpx: 0.25.0
huggingface-hub: 0.18.0
importlib-resources: 6.1.0
jinja2: 3.1.2
markupsafe: 2.1.1
matplotlib: 3.7.0
numpy: 1.23.5
orjson: 3.9.10
packaging: 22.0
pandas: 1.5.3
pillow: 9.4.0
pydantic: 2.4.2
pydub: 0.25.1
python-multipart: 0.0.6
pyyaml: 6.0
requests: 2.28.1
semantic-version: 2.10.0
tomlkit==0.12.0 is not installed.
typer: 0.9.0
typing-extensions: 4.8.0
uvicorn: 0.23.2
websockets: 11.0.3
authlib; extra == 'oauth' is not installed.
itsdangerous; extra == 'oauth' is not installed.


gradio_client dependencies in your environment:

fsspec: 2023.10.0
httpx: 0.25.0
huggingface-hub: 0.18.0
packaging: 22.0
requests: 2.28.1
typing-extensions: 4.8.0
websockets: 11.0.3
```

### Severity

I can work around it

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Adds a `gr.NO_RELOAD` symbol. Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded. Helpful for importing modules that do not like to be reloaded (`tiktoken`, `numpy`) as well as database connections and long running set up code.

Sample usage:

```python
import gradio as gr
import os

if gr.NO_RELOAD:
    print("IMPORTING")
    import tiktoken
    import numpy as np

def image_mod(image):
    print(np.zeros(5))
    return image.rotate(45)


cheetah = os.path.join(os.path.dirname(__file__), "images/lion.jpg")

demo = gr.Interface(image_mod, gr.Image(type="pil", value=cheetah, label="Input"), "image",
    flagging_options=["blurry", "incorrect", "other"], title="Testing a title!", examples=[
        os.path.join(os.path.dirname(__file__), "images/lion.jpg"),
        os.path.join(os.path.dirname(__file__), "images/cheetah1.jpg")
        ])

if __name__ == "__main__":
    demo.launch()
```

Attached gif below, note that the "IMPORTING" statement is not printed during reloads (only when first running the file as expected). 

reload_no_import

Also tested:
- modifying a file imported in the main demo file. Works as expected.
- `gradio cc dev`. Works as expected
- Also adds a reload mode test

Closes: #5402 
Closes: #7123 
Closes #6214

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/cli/commands/reload.py [hunk 1]
```diff
@@ -96,10 +100,23 @@ def _setup_config(
...
+    # copy the watched directories to temp directories
+    # so that we can reload the modules after parsing out
+    # the gr.no_reload contexts
+    for s in watching_dirs:
+        temp_dir = tempfile.mkdtemp()
+        for file in Path(s).rglob("*.py"):
+            src = file
+            dst = Path(temp_dir) / file.relative_to(s)
+            dst.parent.mkdir(parents=True, exist_ok=True)
+            shutil.copy(src, dst)
```
Comment (akx) [inline gradio/cli/commands/reload.py:114]: This will lead to interesting performance characteristics if the the tempdir ends up on a different file system, I'm sure.

Comment (freddyaboulton) [reply gradio/cli/commands/reload.py]: I realized we don't actually need to do the copying

**Comment by abidlabs** [comment]
Would be good add gr.NO_RELOAD here: https://www.gradio.app/guides/developing-faster-with-reload-mode

**Comment by freddyaboulton** [comment]
Open for review - can wait until the release goes out!

File: gradio/utils.py [hunk 2]
```diff
@@ -144,6 +150,51 @@ def swap_blocks(self, demo: Blocks):
...
+        )
+
+    # Find the positions of the code blocks to load
+    for node in ast.walk(tree):
+        if _is_gr_no_reload(node):
+            assert isinstance(node, ast.If)  # noqa: S101
+            node.body = [ast.Pass()]
+
+    # convert tree to string
+    code_removed = ast.unparse(tree)
```
Comment (abidlabs) [inline gradio/utils.py:185]: It looks like `ast.unparse` was added in Python 3.9: https://docs.python.org/3/library/ast.html#ast.unparse

If we absolutely need this method, we could use this third-party implementation: https://github.com/simonpercivall/astunparse

Comment (abidlabs) [reply gradio/utils.py:185]: Btw we should confirm that the reload tests are using 3.8 (otherwise, should not have passed)

Comment (freddyaboulton) [reply gradio/utils.py:185]: Only 7 months until 3.8 EOL 🤩

I think we can manually modify the source code string so will look into it

Comment (abidlabs) [reply gradio/utils.py:185]: Thanks @freddyaboulton, ping me whenever I can take another pass at the pr

Comment (freddyaboulton) [reply gradio/utils.py]: @abidlabs this should be good for review now

File: demo/calculator/run.py
Comment (abidlabs) [inline demo/calculator/run.py:1]: Let's revert the changes in this file since we use this demo in many guides

File: gradio/networking.py
Comment (abidlabs) [inline gradio/networking.py:1]: What changed in this file?

File: guides/09_other-tutorials/developing-faster-with-reload-mode.md [hunk 3]
```diff
@@ -101,6 +101,31 @@ Which you could run like this: `gradio run.py --name Gretel`
 
 As a small aside, this auto-reloading happens if you change your `run.py` source code or the Gradio source code. Meaning that this can be useful if you decide to [contribute to Gradio itself](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) ✅
 
+
+## Controlling the Reload 🎛️
+
+By default, reload mode will re-run your entire script for every change you make.
+But there are some cases where this is not desireable.
```
Comment (abidlabs) [inline guides/09_other-tutorials/developing-faster-with-reload-mode.md:108]:
```suggestion
But there are some cases where this is not desirable.
```

File: js/_website/src/routes/[[version]]/docs/no-reload/+page.server.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	let headers = [
+		["Description", "description"],
+		["Example Uage", "example-usage"]
+	];
+	let method_headers: string[][] = [];
+	const get_slug = make_slug_processor();
+	let obj = {
+		name: "NO_RELOAD",
+		description:
+			"When launching Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded. This is helpful for importing modules that do not like to be reloaded (tiktoken, numpy) as well as database connections and long running set up code.",
```
Comment (abidlabs) [inline js/_website/src/routes/[[version]]/docs/no-reload/+page.server.ts:37]:
```suggestion
			"Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded with gradio's hot reload mode. This is helpful for importing modules that do not like to be reloaded (tiktoken, numpy) as well as database connections and long running set up code.",
```

File: js/_website/src/routes/[[version]]/docs/no-reload/+page.server.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		example: `if gr.NO_RELOAD:
```
Comment (abidlabs) [inline js/_website/src/routes/[[version]]/docs/no-reload/+page.server.ts:38]: I would include a more complete example of a gradio app here to help distinguish between what should be in `if gr.NO_RELOAD` and what shouldn't

File: gradio/state_holder.py [hunk 6]
```diff
@@ -19,6 +19,10 @@ def set_blocks(self, blocks: Blocks):
         self.blocks = blocks
         self.capacity = blocks.state_session_capacity
 
+    def reset(self, blocks: Blocks):
```
Comment (abidlabs) [inline gradio/state_holder.py:22]: Just wondering why the 2nd argument is needed? Why do you need to call `self.set_blocks(blocks)` again?

File: guides/09_other-tutorials/developing-faster-with-reload-mode.md [hunk 7]
```diff
@@ -101,6 +101,31 @@ Which you could run like this: `gradio run.py --name Gretel`
...
+For example, loading a machine learning model should probably only happen once to save time. There are also some python libraries that use C or Rust extensions that throw errors when they are reloaded, like `numpy` and `tiktoken`.
```
Comment (abidlabs) [inline guides/09_other-tutorials/developing-faster-with-reload-mode.md:109]:
```suggestion
For example, loading a machine learning model should probably only happen once to save time. There are also some Python libraries that use C or Rust extensions that throw errors when they are reloaded, like `numpy` and `tiktoken`.
```

**Comment by abidlabs** [comment]
Very cool @freddyaboulton and easy to use!

As I was testing, I noticed two things:

1. Now, when I run a file using `gradio app.py`, the entire contents of the file get executed twice initially. To repro, create a file called `app.py` with these contents:

```py
import gradio as gr
import numpy as np

print("2") 

if gr.NO_RELOAD:
    print("1")

demo = gr.Interface(
    lambda s: s, gr.Image(np.random.random((255, 255, 3))), gr.Image()
)

if __name__ == "__main__":
    demo.launch()
```

You'll see that this is printed to the terminal:
2
1
2
1

2. I had expected that `if gr.NO_RELOAD:` is present in a file that is imported from the main file, it would also work -- since not all heavy imports are necessarily included in the main file. However, that does not seem to be the case. For example, if you have a second file called `run.py` and you import it from `app.py` with:

```
import gradio as gr

if gr.NO_RELOAD:
    print("3")
```

You'll noticed that "3" will be printed to the terminal every time that the original file is changed.

File: gradio/state_holder.py (hunk 6 above)
Comment (freddyaboulton) [reply gradio/state_holder.py:22]: The ids of the second Blocks will be different from the first one

File: gradio/networking.py
Comment (freddyaboulton) [reply gradio/networking.py:1]: I added a new env variable at the top and the whole thing got shifted by one line. There are some minor changes to the `if GRADIO_WATCH_DIRS:` codeblock of `start_server` (basically just passing this new env variable).

**Comment by freddyaboulton** [comment]
Thanks for the review @abidlabs ! I've addressed all of the comments you left in the PR. For point 1, that's expected. In order to reload the modules in the thread running the reloader, the modules need to be first imported. Point 2 should be fixed now!

File: gradio/state_holder.py (hunk 6 above)
Comment (abidlabs) [reply gradio/state_holder.py:22]: Got it

**Review by abidlabs — APPROVED** [review]
Good stuff @freddyaboulton! Love the fact that we now have a test for reload mode

**Comment by freddyaboulton** [comment]
Thanks for the review @abidlabs ! Excited to release this
